FFmpeg: Bump to 4.2.2-Matrix-Alpha1

lrusak opened the pull request:
It's time to get this going. We are behind on ffmpeg and should try to stay as up to date as possible.

The version here is the latest commit in the `release/4.2` branch https://github.com/xbmc/FFmpeg/commits/release/4.2-kodi

I bumped the av libs versions to that provided by the pkgconfig version

```
libavcodec.pc: Version: 58.54.100
libavdevice.pc: Version: 58.8.100
libavfilter.pc: Version: 7.57.100
libavformat.pc: Version: 58.29.100
libavutil.pc: Version: 56.31.100
libpostproc.pc: Version: 55.5.100
libswresample.pc: Version: 3.5.100
libswscale.pc :Version: 5.5.100
```

This will need testing on every platform
 - [x] Android [arm apk](https://mirrors.kodi.tv/test-builds/android/arm/kodi-20200131-925623c0-PR17300-merge-armeabi-v7a.apk) | [arm64 apk](https://mirrors.kodi.tv/test-builds/android/arm64-v8a/kodi-20200131-925623c0-PR17300-merge-arm64-v8a.apk)
 - [ ] iOS [arm64 deb](https://mirrors.kodi.tv/test-builds/darwin/ios-arm64/kodi-20200131-925623c0-PR17300-merge-ios64.deb)
 - [x] TVOS
 - [x] Linux X11
 - [x] Linux GBM
 - [ ] OSX [x86_64 build](https://mirrors.kodi.tv/test-builds/osx/x86_64/kodi-20200131-925623c0-PR17300-merge-x86_64.dmg)
 - [x] Raspberry-Pi
 - [x] Windows [x86_64 exe](https://mirrors.kodi.tv/test-builds/windows/win64/KodiSetup-20200131-925623c0-PR17300-merge-x64.exe) | [x86_64 pdb](https://mirrors.kodi.tv/test-builds/windows/win64/KodiSetup-20200131-925623c0-PR17300-merge-x64.pdb)
 - [ ] Windows UWP


EDIT:
Seems to work fine on linux-gbm with vaapi - http://ix.io/28LC

----

# WARNING

## DON'T MERGE BEFORE TAGGING THE NEW FFMPEG VERSION

lrusak committed:
FFmpeg: Bump to 4.2.2-Matrix-Alpha1

phunkyfish commented:
How come this was never upstreamed?

https://github.com/xbmc/FFmpeg/commit/cfeac74565621ab0fa16395f0184e0d316257900

a1rwulf commented:
> How come this was never upstreamed?
> 
> [xbmc/FFmpeg@cfeac74](https://github.com/xbmc/FFmpeg/commit/cfeac74565621ab0fa16395f0184e0d316257900)

Looks like a workaround for a driver issue, possibly not accepted by ffmpeg because it should get fixed in the driver instead.
@fritsch may know more details about it.

Rechi commented:
Please refresh the FFmpeg patches at `tools/buildsteps/windows/patches`.

phunkyfish commented:
When working on `inputstream.ffmpegdirect` we also added 2 more patches. I don't know if they are relevant the internal kodi implementation. @AlwinEsch ?

This PR also includes the patches kodi uses refreshed to n4.2.2.

fritsch commented:
ffmpeg does not consider this workaround a use-case, especially not one that should be used by default, as it preserves oldish and not so oldish amd cards from crashing we keep it.

lrusak commented:
I had to add libva.pc to the makefile for the `linux-system-libs`

jenkins seems to fail with 
```
ERROR: vaapi requested but not found

If you think configure made a mistake, make sure you are using the latest
version from Git.  If the latest version fails, report the problem to the
ffmpeg-user@ffmpeg.org mailing list or IRC #ffmpeg on irc.freenode.net.
Include the log file "ffbuild/config.log" produced by configure as this will help
solve the problem.
Makefile:78: recipe for target 'x86_64-linux-gnu-debug' failed
make[2]: *** [x86_64-linux-gnu-debug] Error 1
make[2]: Leaving directory '/home/jenkins/workspace/LINUX-64/tools/depends/target/ffmpeg'
Makefile:128: recipe for target 'ffmpeg' failed
```

checking the ffbuild/config.log shows:
```
check_pkg_config vaapi libva >= 0.35.0 va/va.h vaInitialize
test_pkg_config vaapi libva >= 0.35.0 va/va.h vaInitialize
/tmp/tmp.uGCJ0eEXmq/x86_64-linux-gnu-native/bin/pkg-config --exists --print-errors libva >= 0.35.0
Package libva was not found in the pkg-config search path.
Perhaps you should add the directory containing `libva.pc'
to the PKG_CONFIG_PATH environment variable
No package 'libva' found
```

AlwinEsch commented:
@phunkyfish you mean the last three patches? 

| Patch: | Description:
|-|-
| 05-ssl-support-both-pre-1.1.0-and-post-1.1.0-init.patch | to allow openssl "as/or higher as" 1.1.0
| 06-dont-use-libcrypto-locking-functions.patch | I found during search on google, where prevent problems with newer ssl
| 07-ffmpeg-detect-openssl.patch | By static lib build has it not worked on all OS without

phunkyfish commented:
It was 05 and 06. Which wouldn’t apply in kodi core as it uses a set version of OpenSSL. Thanks for explaining.

lrusak commented:
Builds posted for android, ios, osx, and windows, please test!

CastagnaIT commented:
I have ran some videos tests on Windows 10 x64,

play for a few minutes, then skipped about in half, pause and again play
comparing with the branch "master" version i found no difference or problems, also the logs are same

side note:
in my system HEVC 4K is played in software mode (as in master ver branch) i think my video hardware limitation
in GOT_AVC_1080P there aresome logs as:

>nal_unit_type: 1(Coded slice of a non-IDR picture), nal_ref_idc: 0

but i think depend in how the file video was encoded (also happens on master ver branch)

I have the same problems with audio of the master branch version:
I have no audio at all by selecting output device to "direct sound default",
and if i select WASAPI default, the audio working but the volume control no longer works

[LOGs.zip](https://github.com/xbmc/xbmc/files/4164048/LOGs.zip)

lrusak commented:
@CastagnaIT thanks for testing! I'd recommend asking a windows dev or opening an issue report for your other issues 😃

fuzzard commented:
tvos seems ok. Tried a bunch of random media, all played fine. Seeking and chapters seemed all ok.

Thumbnails all worked ok

lrusak commented:
@FernetMenta I added commit https://github.com/xbmc/xbmc/pull/17300/commits/fb1863fb4c98497507367fdaf33fc8829f6c715b to fix an issue we experienced when using the `atempo` filter after https://github.com/FFmpeg/FFmpeg/commit/181031906e4984c1c539dc84d21386a4397e257e. Would you mind taking a look if this is the correct fix?

phunkyfish commented:
I potentially have one more fix. Should know later today.

phunkyfish commented:
@lrusak could you add this patch?

https://github.com/xbmc/inputstream.ffmpegdirect/blob/bd0b7661fd1d5cc7ac54f44a20830c42d15df937/depends/common/ffmpeg/06-ffmpeg-rtp-sources-missing.patch

Confirmed to work via ffmpegdirect. Will get a user to do a quick test once it's there. Ty

lrusak commented:
@phunkyfish I'd rather add the patch to the branch here https://github.com/xbmc/FFmpeg/tree/release/4.2-kodi

I'm think it can be PR'd to that repo and then the I can bump the version.

phunkyfish commented:
It's a bug in ffmpeg. PR created.

lrusak commented:
@phunkyfish I merged it to ffmpeg and bumped the rev here. Is it possible to test somehow?

phunkyfish commented:
I just need a win64 testbuild. As it’s rtp it’s needs to be tested by the reporting users in Germany.

Can I just create one from your branch?

FernetMenta commented:
> to fix an issue we experienced when using the

@lrusak hard to say without knowing what the issue was

btw: are you going back to the old times patching ffmpeg without submitting your patches upstream?

lrusak commented:
> > to fix an issue we experienced when using the
> 
> @lrusak hard to say without knowing what the issue was

@FernetMenta The issue was that the pts in the atempo filter was uninitialized so we would recieve `AV_NOPTS_VALUE`, this caused audioengine to trip up. @peak3d found the issue and can probably explain it better.

> btw: are you going back to the old times patching ffmpeg without submitting your patches upstream?

No, patches should always go upstream.

peak3d commented:
@FernetMenta issue starts at this line: https://github.com/xbmc/xbmc/blob/c3baf14821befc81925613b232ccdcec43d189f2/xbmc/cores/AudioEngine/Engines/ActiveAE/ActiveAEFilter.cpp#L272

which is without the fix AV_NOPTS_VALUE + outsamples according to ffmpeg code.
This leads to a wrong delay and with that wrong cache state (enormous high) and this to stop feeding the audio pipeline (because audio cache is treated as "full".

FernetMenta commented:
@lrusak @peak3d the change in AEFilter should work for this version of ffmpeg but I think it is likely to break in future versions. I expect that the mentioned ffmpeg commit will get revisited because it is inaccurate. You can't calculate future pts from only from start pts. The more you go ahead, the more off pts will get.

Setting pts to zero for every frame seems wrong. If ffmpeg starts to consider not only the first pts, this will break. I suggest that you set pts to zero only for first packet that gets submitted to the filter.

peak3d committed:
[ffmpeg] CActiveAEFilter: initialize frame pts only once after filter creation

This fixes an issue when using the atempo filter after ffmpeg commit 181031906e4984c1c539dc84d21386a4397e257e

lrusak commented:
@FernetMenta updated

FernetMenta commented on xbmc/cores/AudioEngine/Engines/ActiveAE/ActiveAEFilter.cpp:
```diff
@@ -157,6 +157,7 @@ bool CActiveAEFilter::CreateAtempoFilter()
   m_needData = true;
   m_filterEof = false;
   m_started = false;
+  m_ptsInitialized = false;
```
you need this in CloseFilter too

AlwinEsch commented:
It has broken a bit the build on my Ubuntu 19.10 system (by this way https://github.com/xbmc/xbmc/blob/master/docs/README.Linux.md).

It brought:
```
[100%] Linking CXX executable kodi-x11
/usr/bin/ld.gold: Fehler: kann -lunistring nicht finden
collect2: error: ld returned 1 exit status
make[2]: *** [CMakeFiles/kodi.dir/build.make:511: kodi-x11] Fehler 1
make[1]: *** [CMakeFiles/Makefile2:1008: CMakeFiles/kodi.dir/all] Fehler 2
make: *** [Makefile:141: all] Fehler 2
```

After the add with `sudo apt install libunistring-dev` it has worked. Crazy was that the compile has worked, no idea from where he taken then https://packages.ubuntu.com/disco/libunistring-dev to have headers (if they are used).

AlwinEsch commented:
Is that just a coincidence or should `libunistring` also be in`./tools/depends/target`?
